src/irmin/lru.ml

```diff
@@ -79,36 +80,47 @@ module Make (H : Hashtbl.HashedType) = struct
     q : (key * 'a) Q.t;
     mutable cap : int;
     mutable w : int;
+    weight : 'a -> int;
   }
 
-  let weight t = t.w
-  let create cap = { cap; w = 0; ht = HT.create cap; q = Q.create () }
+  let create ?(weight = function _ -> 1) cap =
+    { cap; w = 0; ht = HT.create cap; q = Q.create (); weight }
 
   let drop_lru t =
     match t.q.first with
     | None -> ()
-    | Some ({ Q.value = k, _; _ } as n) ->
-        t.w <- t.w - 1;
+    | Some ({ Q.value = k, v; _ } as n) ->
+        t.w <- t.w - t.weight v;
         HT.remove t.ht k;
         Q.detach t.q n
 
   let remove t k =
     try
       let n = HT.find t.ht k in
-      t.w <- t.w - 1;
+      t.w <- t.w - t.weight (snd n.value);
       HT.remove t.ht k;
       Q.detach t.q n
     with Not_found -> ()
 
+  let clear t =
+    t.w <- 0;
+    HT.clear t.ht;
+    Q.clear t.q
+
   let add t k v =
     if t.cap = 0 then ()
     else (
       remove t k;
       let n = Q.node (k, v) in
-      t.w <- t.w + 1;
-      if weight t > t.cap then drop_lru t;
-      HT.add t.ht k n;
-      Q.append t.q n)
+      let w = t.weight v in
+      if w > t.cap then clear t
```

Comment:
maybe we don't want to clear the LRU here - to avoid big values clearing the cache.

Reply:
I suggest that we don't add the entry to the LRU if its bigger than `cap / K`.

Reply:
what's `K`?

Reply:
Either `1`, meaning that we don't push entries larger than `cap` in the LRU.

Or something like `10`, to reserve the LRU for small entries.